### PR TITLE
Add versioned endpoints

### PR DESCRIPTION
**Pull request by t-persson**
### Applicable Issues

Preparing for: eiffel-community/etos#303

### Description of the Change

Split up ETOS API into v0 and v1alpha

This also has the added effect of separating the API docs to /api/docs, /api/v0/docs and /api/v1alpha/docs.
The default ETOS API version shall be served on /api
Also removed some unecessary code and routers and moved the selftest to /api/etos/ping and /api/version/etos/ping so that we can start doing a few checks when the client pings ETOS.
The old /selftest/ping endpoint still exists, but redirects to the default API version/ping

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: Tobias Persson <tobias.persson@axis.com>

### CODE REVIEW

File: python/src/etos_api/routers/v0/router.py [hunk 1]
```diff
@@ -20,57 +20,76 @@
...
+
+@ETOSv0.get("/selftest/ping")
+async def oldping():
+    """Ping the ETOS service in order to check if it is up and running.
+
+    This is deprecated in favor of `/api/etos/ping`. Implementing here
+    for backward compatibility. In newer API versions this shall not
+    exist.
+    """
+    LOGGER.warning("DEPRECATED request to selftest/ping received!")
```
Comment (andmat900) [inline python/src/etos_api/routers/v0/router.py:91]: No risk for too much logging?

File: python/src/etos_api/routers/v0/router.py [hunk 2]
```diff
@@ -20,57 +20,76 @@
...
-        await SuiteValidator().validate(test_suite)
-    except AssertionError as exception:
-        LOGGER.error("Test suite validation failed!")
-        LOGGER.error(exception)
-        span.add_event("Test suite validation failed")
-        raise HTTPException(
-            status_code=400, detail=f"Test suite validation failed. {exception}"
-        ) from exception
+
+@ETOSv0.get("/ping", tags=["etos"], status_code=204)
```
Comment (andmat900) [inline python/src/etos_api/routers/v0/router.py:73]: why is this changed in v0?

File: python/src/etos_api/routers/v1alpha/utilities.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    :param test_suite_url: The URL to the test suite to validate.
+    """
+    span = trace.get_current_span()
+
+    try:
+        await SuiteValidator().validate(test_suite)
+    except AssertionError as exception:
+        LOGGER.error("Test suite validation failed!")
+        LOGGER.error(exception)
+        span.add_event("Test suite validation failed")
```
Comment (andmat900) [inline python/src/etos_api/routers/v1alpha/utilities.py:105]: Why not record exception as we usually do?

File: python/src/etos_api/routers/v1alpha/utilities.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+def convert_to_rfc1123(value: str) -> str:
+    """Convert string to RFC-1123 accepted string.
+
+    https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names
+
+    Some resource types require their names to follow the DNS label standard as defined in RFC 1123.
+    This means the name must:
+
+        contain at most 63 characters
```
Comment (andmat900) [inline python/src/etos_api/routers/v1alpha/utilities.py:119]: Make sure requirements are checked or enforced by the function.

File: python/src/etos_api/routers/v1alpha/utilities.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+            # This makes the '$regex' query to the event repository more efficient.
+            artifact_identity = f"^{artifact_identity}"
+    else:
+        raise ValueError("'artifact_id' and 'artifact_identity' are both None!")
+    artifact_identifier = artifact_identity or str(artifact_id)
+
+    LOGGER.debug("Wait for artifact created event.")
+    while time.time() < timeout:
+        try:
+            artifacts = await query_handler.execute(query % artifact_identifier)
```
Comment (andmat900) [inline python/src/etos_api/routers/v1alpha/utilities.py:68]: These queries are sometimes very slow. If they are too slow (>30 seconds), there will be no retries as intended.

File: python/src/etos_api/routers/v0/utilities.py [hunk 6]
```diff
@@ -67,3 +73,36 @@ async def wait_for_artifact_created(etos_library, artifact_identity, artifact_id
...
+    """
+    span = trace.get_current_span()
+
+    try:
+        test_suite = await download_suite(test_suite_url)
+        await SuiteValidator().validate(test_suite)
+    except AssertionError as exception:
+        LOGGER.error("Test suite validation failed!")
+        LOGGER.error(exception)
+        span.add_event("Test suite validation failed")
```
Comment (andmat900) [inline python/src/etos_api/routers/v0/utilities.py:105]: Record the exception.

File: python/tests/test_routers.py [hunk 7]
```diff
@@ -95,128 +95,8 @@ def tearDown(self):
         Debug().events_received.clear()
         Debug().events_published.clear()
 
-    def test_head_on_root_without_redirect(self):
```
Comment (andmat900) [inline python/tests/test_routers.py:98]: Add similar v1 tests.

File: python/src/etos_api/routers/v1alpha/utilities.py (hunk 4 above)
Comment (t-persson) [reply python/src/etos_api/routers/v1alpha/utilities.py:119]: Reading just below
`This method does not care about the length of the string since ETOS uses generateName for creating Kubernetes resources and that function will truncate the string down to 63-5 and then add 5 random characters.`

Also, I did not change the code I just moved it.

File: python/src/etos_api/routers/v1alpha/utilities.py (hunk 5 above)
Comment (t-persson) [reply python/src/etos_api/routers/v1alpha/utilities.py:68]: I did not change the code

File: python/src/etos_api/routers/v0/utilities.py (hunk 6 above)
Comment (t-persson) [reply python/src/etos_api/routers/v0/utilities.py:105]: I did not change the code

File: python/src/etos_api/routers/v0/router.py (hunk 1 above)
Comment (t-persson) [reply python/src/etos_api/routers/v0/router.py:91]: I don't think it will be that bad, considering we already have the `INFO:     127.0.0.1:12345 - "GET /selftest/ping HTTP/1.1" 204 No Content` printed every second.
But it's a nice indicator that we should update the ETOS Client to utilize the new endpoint instead (which is something that I will work on soon)

File: python/src/etos_api/routers/v0/router.py (hunk 2 above)
Comment (t-persson) [reply python/src/etos_api/routers/v0/router.py:73]: There is a comment about this on the other function, isnt that sufficient?

File: python/src/etos_api/routers/v0/router.py (hunk 1 above)
Comment (andmat900) [reply python/src/etos_api/routers/v0/router.py:91]: @fredjn Do you see any issues? Should be fine?

File: python/tests/test_routers.py (hunk 7 above)
Comment (andmat900) [reply python/tests/test_routers.py:98]: Then please write an issue about v1 endpoint code not being covered.

File: python/src/etos_api/routers/v1alpha/utilities.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+            assert artifacts["artifactCreated"]["edges"]
+            return artifacts["artifactCreated"]["edges"]
+        except (AssertionError, KeyError):
+            LOGGER.warning("Artifact created not ready yet")
+        await asyncio.sleep(2)
+    LOGGER.error("Artifact %r not found.", artifact_identifier)
+    return None
+
+
+async def download_suite(test_suite_url: str) -> list[dict[str, Any]]:
```
Comment (andmat900) [inline python/src/etos_api/routers/v1alpha/utilities.py:79]: Some of this code seems identical to what is in routers/v0. Cannot it be reused?

File: python/src/etos_api/routers/v0/router.py (hunk 2 above)
Comment (t-persson) [reply python/src/etos_api/routers/v0/router.py:73]: It is implemented in each version because I want it to do verifications as well. v0 does nothing, but I want newer versions to actually check a few things, such as the version of the ETOS client etc.

File: python/src/etos_api/routers/v1alpha/utilities.py (hunk 8 above)
Comment (t-persson) [reply python/src/etos_api/routers/v1alpha/utilities.py:79]: Possibly, but I am not making any big code changes in this PR.

File: python/tests/test_routers.py (hunk 7 above)
Comment (t-persson) [reply python/tests/test_routers.py:98]: Out of scope for this PR. If you wish to have an issue for writing these tests then I think you are better suited to describe what you want :)

File: python/src/etos_api/routers/v0/router.py (hunk 1 above)
Comment (fredjn) [reply python/src/etos_api/routers/v0/router.py:91]: The only issue I'va had with the "I'm fine, I'm fine, ..."-INFO-spamming is that it sometimes forced me to filter it out since i clutters the rest of the logs sometimes. However this warning might be useful to track down clients using old versions of API(s), if we (ever) get OTEL log correlation working it should be fairly trivial, and thus the WARN has a nice overall value.

**Review by andmat900 — APPROVED** [review]

File: python/src/etos_api/routers/v0/utilities.py (hunk 6 above)
Comment (fredjn) [reply python/src/etos_api/routers/v0/utilities.py:105]: IF we do add recording of the exception I suggest removeing the `span.add_event` call, since it then becomes redundant.

File: python/src/etos_api/__init__.py [hunk 9]
```diff
@@ -55,7 +55,7 @@
     trace.set_tracer_provider(PROVIDER)
 
     FastAPIInstrumentor().instrument_app(
-        APP, tracer_provider=PROVIDER, excluded_urls="selftest/.*,logs/.*"
+        APP, tracer_provider=PROVIDER, excluded_urls=".*/etos/ping"
```
Comment (fredjn) [inline python/src/etos_api/__init__.py:58]: Won't this make all the requests to selftest show up in OTEL signals? Sure there is a redirect but the request from the older clients are still to the older urls, or have I got it all backwards?

Comment (t-persson) [reply python/src/etos_api/__init__.py:58]: Update so that the deployment healthcheck uses the new endpoint. The "selftest/ping" endpoint will be used by old clients giving us a heads-up.
